### ISSUE TITLE
Error when loading pickled saved df with version < 0.6.3

### ISSUE THREAD

**Issue by SylvainLan** [geopandas/geopandas#1503]
I'm using geopandas for a while and am used to save my gdf as pickle files, and then open it directly with `pandas`. I have some old files that I want to open. I guess the files were saved as my pandas version was `0.6.2`, but I'm not sure. 

Now that I'm trying to re-open it I get the following error : 
`AttributeError: 'GeoDataFrame' object has no attribute '_item_cache'`

Here is a code sample which makes the problem reproducible : 
with `geopandas==0.6.3` run 
```
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point

data = [[3, 45, 1], [3.1, 45.2, 1.5], [2.7, 44.5, 2.1]]
data = pd.DataFrame(data, columns=['Longitude', 'Latitude', 'value'])
data['geometry'] = list(zip(data['Longitude'], data['Latitude']))
data['geometry'] = data['geometry'].apply(Point)
data = gpd.GeoDataFrame(data)

data.to_pickle('df.pkl')
```

And then with `geopandas==0.8.0` (or `0.7.0`) try : 
```
import pandas as pd
df = pd.read_pickle('df.pkl')
```

It should raise the same error. 

- I'm not sure if it's a problem of `pickle` or `geopandas`, I naively tried to grep `__item_cache` in the gpd code but it wasn't successful either with `0.6.3` or with `0.8.0`
- This is troublesome because I have to keep `0.6.3` in the meantime unless someone sees a turnaround (I could make a "clean save" of all my pickled gdf with `0.6.3` which can be read with `0.8.0` but I don't know if this is possible)
- Maybe my workflow is terrible and then I'll gladly take some advice !

**Comment by martinfleis** [comment]
I am not sure if there is a way how to open old pickle in new pandas. Pickle as a format is not meant for a data storage, especially not with compatibility between versions. The nature of pickle itself does not allow that, if there are significant changes to pandas or geopandas in between the versions. These could be under the hood but still makes the pickle file unusable. My recommendation would be to save those pickle files as some standard geospatial data format with 0.6.3 so you can open that in any other version.

**Comment by SylvainLan** [comment]
I did what you recommended @martinfleis, closing this issue

**Comment by jorisvandenbossche** [comment]
Pickle itself is indeed not necessarily stable across versions, but, when using the pandas `to_pickle`/`read_pickle` methods, pandas has a bunch of compatibility code to ensure reading in pickles from a range of versions is supported.  So this might be a bug in pandas.

@SylvainLan can you still check which versions of pandas you have in both environments?

**Comment by SylvainLan** [comment]
I'm using 1.0.5

**Comment by jorisvandenbossche** [comment]
Also in the environment with geopandas 0.6.3?

**Comment by SylvainLan** [comment]
Yes, when I run `pip install 'geopandas==0.6.3'` I don't get asked to downgrade pandas too

**Comment by SylvainLan** [comment]
I ran the example above with `pandas==1.0.0` or `pandas==0.25.0` and it raises the same error

**Comment by jorisvandenbossche** [comment]
OK, thanks, I can indeed reproduce this with pandas 1.0 and geopandas 0.6 vs 0.8. 
I assume this is due to changes in the CRS attribute handling. It is now a property with a getter/setter, and not a simple attribute on the class. The problem seems to be that while unpickling, it tries to set `crs` which fails if the geodataframe object is not yet fully initialized.

**Comment by jorisvandenbossche** [comment]
The problem comes from the fact that older pandas versions (< 0.7) used `crs` as entry in `_metadata` and now we use `_crs`. So when unpickling, it tries to set `crs` instead of `_crs` (going through the property settter instead of the basic attribute assignment).

We might be able to solve this with some compat code in GeoDataFrame:

```python
    def __setstate__(self, state):
        # overriding DataFrame method for compat with older pickles (CRS handling)

        if isinstance(state, dict):
            if "_metadata" in state:
                metadata = state["_metadata"]
                metadata[metadata.index('crs')] = "_crs"
            if "crs" in state and "_crs" not in state:
                crs = state.pop("crs")
                state["_crs"] = pyproj.CRS.from_user_input(crs) if crs is not None else crs
        
        super().__setstate__(state)
```

this already reads in the pickle file into a GeoDataFrame fine, only it now still creates a GeometryArray without `_crs` .. so some more compat code would be needed there.

**Comment by SylvainLan** [comment]
I tried resolving what you mentioned above @jorisvandenbossche but I couldn't do any better than to set the geometry's crs at the end of the method `self.geometry.crs = self._crs`

**Comment by jorisvandenbossche** [comment]
Closed by #1511
